jumanji/environments/routing/connector/generation_methods.py

```diff
@@ -0,0 +1,437 @@
+
+        Args:
+            grid: the current layout of the board i.e. current grid.
+            cell: the flat index of the cell to find adjacent cells of.
+
+        Returns:
+            A chex.Array of size 4 with the flat indices of adjacent cells.
+        """
+        adjacent_cells = self._adjacent_cells(cell)
+        # Get the wire id of the current cell
+        value = grid[jnp.divmod(cell, self.cols)]
+        wire_id = (value - 1) // 3
+
+        _, available_cells_mask = jax.lax.scan(self._is_cell_free, grid, adjacent_cells)
+        # Also want to check if the cell is touching itself more than once
+        _, touching_cells_mask = jax.lax.scan(
+            self._is_cell_doubling_back, (grid, wire_id), adjacent_cells
+        )
+        available_cells_mask = available_cells_mask & touching_cells_mask
+        available_cells = jnp.where(available_cells_mask == 0, -1, adjacent_cells)
+
+        return available_cells
+
+    def _is_cell_free(
+        self,
+        grid: chex.Array,
+        cell: chex.Array,
+    ) -> Tuple[chex.Array, chex.Array]:
+        """Check if a given cell is free, i.e. has a value of 0.
+
+        Args:
+            grid: the current grid of the board.
+            cell: the flat index of the cell to check.
+
+        Returns:
+            A tuple of the new grid and a boolean indicating whether the cell is free or not.
+        """
+        coordinate = jnp.divmod(cell, self.cols)
+        return grid, jax.lax.select(
+            cell == -1, False, grid[coordinate[0], coordinate[1]] == 0
+        )
+
+    def _is_cell_doubling_back(
+        self,
+        grid_wire_id: Tuple[chex.Array, int],
+        cell: int,
+    ) -> Tuple[Tuple[chex.Array, int], bool]:
+        """Checks if moving into and adjacent position would result in a wire doubling back on itself.
+
+        Check if the cell is touching any of the wire's own cells more than once.
+        This means looking for surrounding cells of value 3 * wire_id + POSITION or
+        3 * wire_id + PATH.
+        """
+        grid, wire_id = grid_wire_id
+        # Get the adjacent cells of the current cell
+        adjacent_cells = self._adjacent_cells(cell)
+
+        def is_cell_doubling_back_inner(
+            grid: chex.Array, cell: chex.Array
+        ) -> Tuple[chex.Array, chex.Array]:
+            coordinate = jnp.divmod(cell, self.cols)
+            cell_value = grid[coordinate[0], coordinate[1]]
+            touching_self = jnp.logical_or(
+                jnp.logical_or(
+                    cell_value == 3 * wire_id + POSITION,
+                    cell_value == 3 * wire_id + PATH,
+                ),
+                cell_value == 3 * wire_id + TARGET,
+            )
+            return grid, jnp.where(cell == -1, False, touching_self)
+
+        # Count the number of adjacent cells with the same wire id
+        _, doubling_back_mask = jax.lax.scan(
+            is_cell_doubling_back_inner, grid, adjacent_cells
+        )
```

Comment:
This can be parallelised as well.
```suggestion
        def is_cell_doubling_back_inner(
            grid: chex.Array, cell: chex.Array
        ) -> chex.Array:
            coordinate = jnp.divmod(cell, self.cols)
            cell_value = grid[tuple(coordinate)]
            touching_self = (cell_value == 3 * wire_id + POSITION) | (cell_value == 3 * wire_id + PATH) | (cell_value == 3 * wire_id + TARGET)
            return (cell != -1) & touching_self

        # Count the number of adjacent cells with the same wire id
        doubling_back_mask = jax.vmap(is_cell_doubling_back_inner, in_axes=(None, 0))(grid, adjacent_cells)
```

Reply:
Done